src/contracts/middleware/BLSRegistryCoordinatorWithIndices.sol

```diff
@@ -0,0 +1,325 @@
+    }
+
+    function _registerOperatorWithCoordinator(address operator, bytes calldata quorumNumbers, BN254.G1Point memory pubkey) internal {
+        require(
+            slasher.contractCanSlashOperatorUntilBlock(operator, address(serviceManager)) == type(uint32).max,
+            "StakeRegistry._registerOperator: operator must be opted into slashing by the serviceManager"
+        );
+        
+        // check that the sender is not already registered
+        require(_operators[operator].status != OperatorStatus.REGISTERED, "BLSIndexRegistryCoordinator._registerOperatorWithCoordinator: operator already registered");
+
+        // get the quorum bitmap from the quorum numbers
+        uint192 quorumBitmap = uint192(BytesArrayBitmaps.orderedBytesArrayToBitmap_Yul(quorumNumbers));
+        require(quorumBitmap != 0, "BLSIndexRegistryCoordinator._registerOperatorWithCoordinator: quorumBitmap cannot be 0");
+
+        // register the operator with the BLSPubkeyRegistry and get the operatorId (in this case, the pubkeyHash) back
+        bytes32 operatorId = blsPubkeyRegistry.registerOperator(operator, quorumNumbers, pubkey);
+
+        // register the operator with the StakeRegistry
+        stakeRegistry.registerOperator(operator, operatorId, quorumNumbers);
+
+        // register the operator with the IndexRegistry
+        indexRegistry.registerOperator(operatorId, quorumNumbers);
+
+        // set the operatorId to quorum bitmap history
+        _operatorIdToQuorumBitmapHistory[operatorId].push(QuorumBitmapUpdate({
+            updateBlockNumber: uint32(block.number),
+            nextUpdateBlockNumber: 0,
+            quorumBitmap: quorumBitmap
+        }));
+
+        // set the operator struct
+        _operators[operator] = Operator({
+            operatorId: operatorId,
+            fromTaskNumber: serviceManager.taskNumber(),
+            status: OperatorStatus.REGISTERED
+        });
+
+        // record a stake update not bonding the operator at all (unbonded at 0), because they haven't served anything yet
+        serviceManager.recordFirstStakeUpdate(operator, 0);
+    }
+
+    function _deregisterOperatorWithCoordinator(address operator, bytes calldata quorumNumbers, BN254.G1Point memory pubkey, bytes32[] memory operatorIdsToSwap, uint32 globalOperatorListIndex) internal {
+        require(_operators[operator].status == OperatorStatus.REGISTERED, "BLSIndexRegistryCoordinator._deregisterOperatorWithCoordinator: operator is not registered");
+
+        // get the operatorId of the operator
+        bytes32 operatorId = _operators[operator].operatorId;
+        require(operatorId == pubkey.hashG1Point(), "BLSIndexRegistryCoordinator._deregisterOperatorWithCoordinator: operatorId does not match pubkey hash");
+
+        // get the quorumNumbers of the operator
+        uint192 quorumsToRemoveBitmap = uint192(BytesArrayBitmaps.orderedBytesArrayToBitmap_Yul(quorumNumbers));
```

Comment:
just noting that we similarly need to check the `quorumNumbers` input here (should verify that the library output is <= type(uint192).max) -- we could add a library function for this, perhaps?

Reply:
discussed offline -- seems like it could make sense to have a library function with a uint8 input `maxIndex` param which calculates the bitmap and then checks something like `resultingBitmap < (1 << maxIndex)`

Reply:
https://github.com/Layr-Labs/eigenlayer-contracts/pull/76/commits/2d4a033ef22fdb60df4b1bdb26907986e3f6a44a